server/src/main/java/org/opensearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -694,12 +701,95 @@ public RetryOnReplicaException(StreamInput in) throws IOException {
         }
     }
 
+    private class ReplicaActionFactory {
+
+        private ReplicaAction create(
+            IndexShard replica,
+            ReplicationMode replicationMode,
+            ConcreteReplicaRequest<ReplicaRequest> replicaRequest,
+            ActionListener<ReplicaResponse> onCompletionListener,
+            ReplicationTask task
+        ) {
+            if (replicationMode == ReplicationMode.FULL_REPLICATION) {
+                return new FullReplicationAsyncReplicaAction(replicaRequest, onCompletionListener, task, replica);
+            } else if (replicationMode == ReplicationMode.PRIMARY_TERM_VALIDATION) {
+                if (replica.isRemoteTranslogEnabled()) {
+                    return new FullReplicationAsyncReplicaAction(replicaRequest, onCompletionListener, task, replica);
+                } else {
+                    return new PrimaryTermValidationReplicaAction(replicaRequest, onCompletionListener, task, replica);
+                }
+            }
+            setPhase(task, "failed");
+            throw new UnsupportedReplicaActionException(
+                "ReplicationMode:{} is not supported on targetAllocationId={} for shard={}",
+                replicationMode,
+                replicaRequest.getTargetAllocationID(),
+                replica.shardId()
+            );
```

Comment:
I think we need to invoke `onCompletionListener.onFailure()` I'll prefer an existing exception or IllegalStateException to avoid unnecessary serialisation bugs

Reply:
The alternate being here is throw `RuntimeException` . Then this failure handling can be done below

https://github.com/opensearch-project/OpenSearch/blob/00e4e9225c2d4a8b4436d119115f9214f42ca587/server/src/main/java/org/opensearch/action/support/replication/TransportReplicationAction.java#L671-L672